astropy/coordinates/builtin_frames/skyoffset.py

```diff
@@ -133,20 +118,57 @@ class SkyOffsetFrame(BaseCoordinateFrame):
     rotation = QuantityAttribute(default=0, unit=u.deg)
     origin = CoordinateAttribute(default=None, frame=None)
 
-    def __new__(cls, *args, **kwargs):
-        # We don't want to call this method if we've already set up
-        # an skyoffset frame for this class.
-        if not (issubclass(cls, SkyOffsetFrame) and cls is not SkyOffsetFrame):
-            # We get the origin argument, and handle it here.
-            try:
-                origin_frame = kwargs['origin']
-            except KeyError:
-                raise TypeError("Can't initialize an SkyOffsetFrame without origin= keyword.")
-            if hasattr(origin_frame, 'frame'):
-                origin_frame = origin_frame.frame
-            newcls = make_skyoffset_cls(origin_frame.__class__)
-            return newcls.__new__(newcls, *args, **kwargs)
+    # ---------------------------------------------------------------
+    # `astropy.utils.metaclasses.InheritanceInMixMeta` customizations
 
+    @classmethod
+    def _inmix_prepare_type(cls, framecls, base_cls):
+        name = 'SkyOffset' + framecls.__name__
+        bases = (base_cls, framecls)
+        namespace = {
+            'origin': CoordinateAttribute(frame=framecls, default=None),
+            # The following two have to be done because otherwise we use the
+            # defaults of SkyOffsetFrame set by BaseCoordinateFrame.
+            '_default_representation': framecls._default_representation,
+            '_default_differential': framecls._default_differential,
+            '__doc__': SkyOffsetFrame.__doc__,
+         }
+
+        return name, bases, namespace
+
+    @classmethod
+    def _inmix_make_class(cls, framecls):
+        if not issubclass(framecls, BaseCoordinateFrame):
+            raise TypeError
+
+        # Call super
+        skyoffset_framecls = type(cls)._inmix_make_class(cls, framecls)
+
+        # Register transformations
+        frame_transform_graph.transform(FunctionTransform, skyoffset_framecls, skyoffset_framecls)(_skyoffset_to_skyoffset)
+        frame_transform_graph.transform(DynamicMatrixTransform, framecls, skyoffset_framecls)(_reference_to_skyoffset)
+        frame_transform_graph.transform(DynamicMatrixTransform, skyoffset_framecls, framecls)(_skyoffset_to_reference)
+
+        # TODO! deprecate _skyoffset_cache
+        _skyoffset_cache[framecls] = skyoffset_framecls
+
+        return skyoffset_framecls
+
+    @classmethod
+    def _inmix_make_instance(cls, *args, **kwargs):
```

Comment:
I'm finding that the logic of the metaclass' methods are a bit complicated. Having small functions that the child metaclasses override keeps things simpler. Especially since most child metaclasses just need to make one or two changes then call ``super``.